File: util/alt_lang/ann.py
"""Basic implementation of an artificial neural network in python using a
sigmoid activation function and the backpropagation learning algorithm.

The network attemps to learn the behaviour of an XOR gate.

The theory used for the implementation is from:
"Machine Learning" by Tom M. Mitchell (ISBN: 0-07-115467-1)

"""
import math
import numpy as np
import random


class ANN(object):
    """Implementation of an artificial neural network using a sigmoid 
    activation function and the backpropagation learning algorithm.
    """

    def __init__(self, topology, learning_rate=0.05):
        """Construct an ANN containing a specific number of input, hidden and
        output nodes.

        Multi-dimensional arrays are defined to hold the network outputs (o),
        weights (w) and error terms (e).

        Args:
            topology: ([int]) Decription of the topology of the network.
                
                Eg: [2, 2, 1] describes a network with 2 input nodes, 2 hidden
                nodes and 1 output node.

            learning_rate: (float) The learning rate of the backpropagation
                algorith.
        """

        self._o = map(lambda n: np.zeros(n), topology)
        self._e = map(lambda n: np.zeros(n), topology)

        self._w = []
        for l in range(1, len(topology)):
            def weights_for_node(_):
                return map(lambda _: random.random()-.5, range(topology[l-1]))
            layer = map(weights_for_node, range(topology[l]))
            self._w.append(layer)
        self._w.insert(0, None) # The input layer doesn't have weights.

        self._topology = topology
        self._learning_rate = learning_rate

    def train(self, training_set):
        """Train the network using a complete training set.

        Args:
            training_set: ([(input_vector, target_ouput)]) Training set.

        Returns:
            (float) Accumulated error for complete training set.

        """
        e = 0.0
        for input_vector, target_output in training_set:
            o = self.evaluate(input_vector)
            e += self._error(o, target_output)
            self._learn(target_output)
        return e

    def evaluate(self, xs):
        """Return the output vector for an input vector, given the current
        weights.

        First the input vector is loaded into the input nodes, then moving
        forward through the network update the output value of each node.
        Finally the output values belonging to the final layer are returned
        as the output vector.

        Args:
            xs: (list) Input vector.

        Returns:
            (list) Output vector.

        See: p. 97 (4.12)
        """
  
        self._o[0] = np.array(xs)

        for l in range(1, len(self._topology)):
            for i in range(self._topology[l]):
                o = np.dot(self._w[l][i], self._o[l-1])
                self._o[l][i] = self._sigmoid(o)

        return self._o[-1]
    
    def _learn(self, target_outputs):
        """Optionally called after 'evaluate' (once the outputs for an input
        vector have been set). The weights are adjusted to better produce
        the 'target_outputs' on classifying the same input vector again.

        Args:
            target_outputs: ([float]) Target outputs.

        See: p. 98 (T4.5)
        """

        # Calculate the error terms for the output nodes by comparing the 
        # target output with the actual output.
        for i,t in enumerate(target_outputs):
            o = self._o[-1][i]
            self._e[-1][i] = o * (1-o) * (t-o)

        # Calculate the error terms for all hidden nodes by moving backwards
        # through the network (starting with the layer just before the output
        # layer) and using the dot product of the weights from the next level
        # that connect upstream to the node and the error terms from the next 
        # level.
        for l in range(len(self._topology)-2, 0, -1):
            for i in range(self._topology[l]):
                w = map(
                    lambda j: self._w[l+1][j][i], range(self._topology[l+1]))
                e = np.dot(w, self._e[l+1])
                o = self._o[l][i]
                self._e[l][i] = o * (1-o) * e

        # Update weights by moving forward through the network (starting with
        # the second layer), calculating the weight delta for a given node by
        # multiplying the weight between the node and each upstream node by the
        # current output value from each upstream node (and the learning rate).
        for l in range(1, len(self._topology)):
            for i in range(self._topology[l]):
                for j in range(self._topology[l-1]):
                    delta = \
                        self._learning_rate * self._e[l][i] * self._o[l-1][j]
                    self._w[l][i][j] += delta

    def _sigmoid(self, x):
        """The 'squashing function' that produces an S-shaped output from the
        activation functioning allowing the network to model complexity beyond
        liner separability.

        Args:
            x: (float) Output from activation function before being 'squashed'.

        Returns:
            (float) 'Squashed' x.

        See: p. 97 (4.12)
        """
        return 1 / (1 + math.exp(-x))

    def _error(self, actual_output, target_output):
        """Return the error value between the actual and target outputs for a
        single training example.

        Args:
            actual_output: ([float]) Actual output.
            target_output: ([float]) Target output.

        Returns:
            (float) Error value.

        See: p. 97 (4.13) (modified)
        """
        e = 0
        for t,o in zip(target_output, actual_output):
            e += (t-o)**2
        return e * 0.5


if __name__ == "__main__":

    # Initialise the artificial neural network.
    training_set = [
        ([0, 0], [0]),
        ([1, 0], [1]),
        ([0, 1], [1]),
        ([1, 1], [0]),
        ]
    ann = ANN(
        topology=[2, 2, 1],
        learning_rate=0.05)

    # Iteratively learn from the training set until an acceptable error value 
    # is obtained.
    e = float('inf')
    i = 0
    while e > .2:
        e = ann.train(training_set)
        if i % 100 == 0:
            print "Error %s" % e
        i += 1

    # Display learnt XOR classification.
    for x, _ in training_set:
        print "%s -> %s" % (x, ann.evaluate(x))

File: util/alt_lang/ann.hs
import Data.Array
import Data.List
import Debug.Trace
import System.Random
import System.Environment
import Data.Maybe

type InputVector =          [Float]
type TargetOutputVector =   [Float]
type ActualOutputVector =   [Float]
type Outputs =              Array (Int, Int) Float
type ErrorTerms =           Array (Int, Int) Float
type Weights =              Array (Int, Int, Int) Float
type TrainingSet =          [([Float], [Float])]
data Network =              Network Outputs ErrorTerms Weights

-- Config ----------------------------------------------------------------------

-- topology = [10, 15, 10, 6] -- binadd
topology = [784, 1176, 392, 10] -- digit
threshold = 0.1
learningRate = 0.5

-- Intialise Network -----------------------------------------------------------

initNetwork :: IO Network
initNetwork = do
    rs' <- rs
    return (Network os es (ws rs'))
    where rs = sequence $ map (\_ -> randomRIO (-0.5, 0.5)) rangeWeights
          os = listArray boundsOutputs [0,0..] 
          es = listArray boundsErrorTerms [0,0..]
          ws rs = listArray boundsWeights [0,0..] // zip rangeWeights rs

-- | Return complete range for all used weight elements.
rangeWeights :: [(Int, Int, Int)]
rangeWeights = concat . map rangeWeights' $ range (1, lastLayer)

-- | Return range for all used weight elements in a layer.
rangeWeights' :: Int -> [(Int, Int, Int)]
rangeWeights' l = concat . map (rangeWeights'' l) $ rangeLevel l

-- | Return range for all used weight elements for a node in a layer.
rangeWeights'' :: Int -> Int -> [(Int, Int, Int)]
rangeWeights'' l i = [(l, i, j) | j <- rangeLevel (l-1)]

-- | Given the topology return the indices available at a level.
rangeLevel :: Int -> [Int]
rangeLevel l = range (0, (topology !! l)-1)

lastLayer =         (length topology) -1
maxLayerWidth =     (maximum topology) -1
boundsOutputs =     ((0, 0),    (lastLayer, maxLayerWidth))
boundsErrorTerms =  ((0, 0),    (lastLayer, maxLayerWidth))
boundsWeights =     ((0, 0, 0), (lastLayer, maxLayerWidth, maxLayerWidth))

-- Network Views ---------------------------------------------------------------

-- | Return the weights on level 'l' for node 'i'.
weightsAt :: Array (Int, Int, Int) Float -> Int -> Int -> [Float]
weightsAt ws l i = map (ws!) (rangeWeights'' l i)

-- Return the outputs at layer 'l'.
outputsAt :: Outputs -> Int -> [Float]
outputsAt os l = [os!(l, i) | i <- rangeLevel l]

-- Return the outputs at layer 'l' given a network.
outputsAt' :: Int -> Network -> [Float]
outputsAt' l (Network os _ _) = outputsAt os l

-- Return the range of output indices at layer 'l'.
rngOutputs :: Int -> [(Int, Int)]
rngOutputs l = [(l, i) | i <- rangeLevel l]

-- Return the range of error term indices at layer 'l'.
rngErrorTerms :: Int -> [(Int, Int)]
rngErrorTerms l = [(l, i) | i <- rangeLevel l]

-- Learning --------------------------------------------------------------------

-- | Given an input vector, move forward through the network setting the output
-- value for each node.
setOutputs :: InputVector -> Network -> Network
setOutputs xs (Network os es ws) = Network os' es ws
    where input =       os // zip (rngOutputs 0) xs
          os' =         foldl' layer input [1..lastLayer]
          layer os l =  os // [((l, i), f os l i) | i <- rangeLevel l] 
          f os l i =    sigmoid $ dot (weightsAt ws l i) (outputsAt os (l-1))

-- | Moving backwards through the network set the error term for each node
-- based on its contribution towards the difference between the actual output
-- and target output vectors.
setErrorTerms :: TargetOutputVector -> Network -> Network
setErrorTerms ts n@(Network os es ws) = Network os es' ws
    where es' = foldl' f output [lastLayer-1,lastLayer-2..1] 
          f es l = es // [((l, i), g es l i) | i <- rangeLevel l]
          g es l i = let e = dot (weights (l+1) i) (errorTerms es (l+1))
                         o = os!(l, i)
                     in o * (1-o) * e
          weights l i =     [ws!(l, j, i) | j <- rangeLevel l]
          errorTerms es l = [es!(l, j) | j <- rangeLevel l]
          output = setOutputErrorTerms ts n

-- | Set the error terms for the output layer.
setOutputErrorTerms :: TargetOutputVector -> Network -> ErrorTerms
setOutputErrorTerms ts (Network os es _) = es'
    where es' = es // zip (rngErrorTerms lastLayer) 
                      (map f (zip (outputsAt os lastLayer) ts)) 
          f (o, t) = o * (1-o) * (t-o)

-- | Adjust the network weights according to a learning rate.
setWeights :: Float -> Network -> Network
setWeights lr (Network os es ws) = Network os es ws'
    where ws' = ws // [(i, f i) | i <- rangeWeights]
          f (l, i, j) = ws!(l, i, j) + (lr * es!(l, i) * os!(l-1, j))

-- | Given a single pair of actual and target output vectors return the 
-- associated error value.
errorVal :: [Float] -> [Float] -> Float -- Error
errorVal os ts = (sum $ zipWith (\t o -> (t-o)**2) ts os) * 0.5

correctClassify :: [Float] -> [Float] -> Int
correctClassify os ts = let os' = map (fromIntegral . round) os
                        in if os' == ts then 1 else 0

-- | Train the network on each member of the training set, then return the
-- final network and its associated error.
train' :: TrainingSet -> Network -> (Network, Float, Int)
train' tset n = foldl' f (n, 0, 0) tset
    where f (n, e, c) (x, t) = let n' = setErrorTerms t $ setOutputs x n
                                   o = outputsAt' lastLayer n'
                                   e' = errorVal o t
                                   c' = correctClassify o t
                                   n'' = setWeights learningRate n'
                               in (n'', e+e', c+c')

-- | Repeatedly train the network until its error, after having processed all
-- members of the training set, is below the threshold.
train :: TrainingSet -> Network -> Int -> IO Network
train tset n i = do
    let (n'@(Network _ _ ws), e, c) = train' tset n
        e' = trace i e ws c
    if e' < threshold || areWeightsSame n n' then return n' else train tset n' (i+1)
    where trace i e ws c
              | rem i 10 == 0 = traceShow (show e ++ "-" ++ show c ++ "/" ++ show (length tset)) e
              | otherwise =     e

areWeightsSame :: Network -> Network -> Bool
areWeightsSame (Network _ _ ws1) (Network _ _ ws2) = ws1 == ws2

-- Helpers ---------------------------------------------------------------------

-- | The Sigmoid function.
sigmoid :: Float -> Float
sigmoid x = 1.0 / (1 + exp (-x))

-- | Dot product of two vectors.
dot :: Num a => [a] -> [a] -> a 
dot xs ys = sum (zipWith (*) xs ys)

-- Main ------------------------------------------------------------------------

test :: TrainingSet -> Network -> IO ()
test tset n = do
    let output = map f tset
    mapM_ putStrLn output
    where f (x, t) = let o = (outputsAt' lastLayer) $ setOutputs x n
                     in (show x ++ "=" ++ show t ++ " => " ++ show o)

trainingSet :: String -> IO TrainingSet
trainingSet fn = do
    xs <- readFile fn
    return . map fmt $ lines xs
    where fmt x = let (a, b) = splitAt (fromJust $ elemIndex ':' x) x
                      a' = map read $ splitOn ',' a
                      b' = map read . splitOn ',' $ tail b
                  in (a', b')

splitOn :: Eq a => a -> [a] -> [[a]]
splitOn d xs = case elemIndex d xs of
                   Nothing ->  [xs]
                   (Just i) -> let (x', xs') = splitAt i xs
                               in x' : splitOn d (tail xs')

main = do
    fn <- fmap head getArgs
    tset <- trainingSet fn
    n <- initNetwork
    n' <- train tset n 0
    test tset n'

